File: libfindchars-examples/src/main/java/zz/customname/FindUtf8Characters.java
package zz.customname;

import java.lang.foreign.MemorySegment;
import java.nio.charset.StandardCharsets;

import org.knownhosts.libfindchars.api.MatchStorage;
import org.knownhosts.libfindchars.generator.Utf8EngineBuilder;

/**
 * Demonstrates detecting multi-byte UTF-8 characters alongside ASCII.
 */
class FindUtf8Characters {

    public static void main(String[] args) {

        var result = Utf8EngineBuilder.builder()
                .codepoints("whitespace", ' ', '\n')
                .codepoint("eacute", 0xE9)       // é  2-byte
                .codepoint("trademark", 0x2122)   // ™  3-byte
                .codepoint("grin", 0x1F600)       // 😀 4-byte
                .build();
        var engine = result.engine();
        var literals = result.literals();

        byte wsLit = literals.get("whitespace");
        byte eLit = literals.get("eacute");
        byte tmLit = literals.get("trademark");
        byte grinLit = literals.get("grin");

        // UTF-8 test string: "café ™ 😀 fin\n"
        String input = "caf\u00E9 \u2122 \uD83D\uDE00 fin\n";
        byte[] utf8 = input.getBytes(StandardCharsets.UTF_8);

        // Pad to at least vector size
        byte[] data = new byte[Math.max(64, utf8.length)];
        System.arraycopy(utf8, 0, data, 0, utf8.length);

        var segment = MemorySegment.ofArray(data);
        var matchStorage = new MatchStorage(64, 32);
        var match = engine.find(segment, matchStorage);

        System.out.println("Input: " + input.trim());
        System.out.println("UTF-8 bytes: " + utf8.length);
        System.out.println("Matches found: " + match.size());

        for (int i = 0; i < match.size(); i++) {
            byte lit = match.getLiteralAt(matchStorage, i);
            int pos = match.getPositionAt(matchStorage, i);

            if (lit == wsLit) {
                System.out.println("  whitespace at byte " + pos);
            } else if (lit == eLit) {
                System.out.println("  é (U+00E9) at byte " + pos);
            } else if (lit == tmLit) {
                System.out.println("  ™ (U+2122) at byte " + pos);
            } else if (lit == grinLit) {
                System.out.println("  😀 (U+1F600) at byte " + pos);
            }
        }
    }
}

File: rust/findchars-examples/examples/find_utf8_characters.rs
//! Detect multi-byte UTF-8 codepoints (2-, 3-, and 4-byte) alongside ASCII.
//!
//! Run with:
//! ```text
//! cargo run --example find_utf8_characters -p findchars-examples
//! ```
//!
//! Rust equivalent of the Java `FindUtf8Characters` example.

use findchars::{EngineBuilder, MatchStorage};

fn main() {
    // `codepoint` accepts any Unicode scalar value; the builder encodes it to
    // UTF-8 and the engine detects the full byte sequence.
    let result = EngineBuilder::new()
        .codepoints("whitespace", b" \n")
        .codepoint("eacute", 0x00E9) // é  — 2-byte: C3 A9
        .codepoint("trademark", 0x2122) // ™  — 3-byte: E2 84 A2
        .codepoint("grin", 0x1F600) // 😀 — 4-byte: F0 9F 98 80
        .build()
        .expect("solver failed");

    let engine = &result.engine;
    let literals = &result.literals;

    let whitespace = literals["whitespace"];
    let eacute = literals["eacute"];
    let trademark = literals["trademark"];
    let grin = literals["grin"];

    // Positions are reported as byte offsets into the UTF-8 stream — a
    // multi-byte match is reported at its lead byte.
    let text = "café ™ 😀 fin\n";
    let data = text.as_bytes();

    let mut storage = MatchStorage::new(64);
    let view = engine.find(data, &mut storage);

    println!("input: {}", text.trim_end());
    println!("utf-8 bytes: {}", data.len());
    println!("matches found: {}", view.len());

    for i in 0..view.len() {
        let pos = view.position(i);
        let lit = view.literal(i);

        let kind = if lit == whitespace {
            "whitespace"
        } else if lit == eacute {
            "é (U+00E9)"
        } else if lit == trademark {
            "™ (U+2122)"
        } else if lit == grin {
            "😀 (U+1F600)"
        } else {
            "unknown"
        };

        println!("  {kind} at byte {pos}");
    }
}
